packages/@aws-cdk/core/lib/cfn-parse.ts

```diff
@@ -441,11 +457,54 @@ export class CfnParser {
 
     const key = objectKeys[0];
     return key === 'Ref' || key.startsWith('Fn::') ||
-        // special intrinsic only available in the 'Conditions' section
-        (this.options.context === CfnParsingContext.CONDITIONS && key === 'Condition')
+      // special intrinsic only available in the 'Conditions' section
+      (this.options.context === CfnParsingContext.CONDITIONS && key === 'Condition')
       ? key
       : undefined;
   }
+
+  private parseFnSubString(value: string, map: any): string {
+    const leftBrace = value.indexOf('${');
+    const rightBrace = value.indexOf('}') + 1;
+    const leftHalf = value.substring(0, leftBrace);
+    const rightHalf = value.substring(rightBrace);
+    // don't include left and right braces when searching for the target of the reference
+    const refTarget = value.substring(leftBrace + 2, rightBrace - 1).trim();
+    if (refTarget === '') {
+      return '';
+    } else if (refTarget[0] === '!') {
+      return value.substring(0, rightBrace) + this.parseFnSubString(rightHalf, map);
+    }
+
+    // lookup in map
+    let refElement;
+    if (map) {
+      refElement = map[refTarget];
+      if (refElement) {
+        return leftHalf + '${' + refTarget + '}' + this.parseFnSubString(rightHalf, map);
+      }
+    }
+
+    // since it's not in the map, check if it's a pseudo parameter
+    const specialRef = specialCaseSubRefs(refTarget);
+    if (specialRef) {
+      return leftHalf + specialRef + this.parseFnSubString(rightHalf, map);
+    }
+
+    // handle Ref
+    refElement = this.options.finder.findRefTarget(refTarget);
+    if (refElement) {
+      return leftHalf + CfnReference.for(refElement, 'Sub', 'Ref') + this.parseFnSubString(rightHalf, map);
+    }
+
+    // handle GetAtt
+    refElement = this.options.finder.findResource(refTarget.substring(0, refTarget.indexOf('.')));
+    if (refElement) {
+      return leftHalf + refElement.getAtt(refTarget.substring(refTarget.indexOf('.') + 1), 'GetAtt') + this.parseFnSubString(rightHalf, map);
+    }
+
+    throw new Error(`Element used in Ref expression with logical ID: '${refTarget}' in Fn::Sub not found`);
```

Comment:
This is the wrong message in the `${A.B}` case though, right?

I think you want something like this:

```ts
const isRef = refTarget.indexOf('.') === -1'
if (isRef) {
  const refElement = this.options.finder.findRefTarget(...
  if (!refElement) throw new Error(`Element referenced in Fn::Sub expression with logical ID ...
  return ...;
} else {
  const refResource = this.options.finder.findResource(...
  if (!refResource) throw new Error(`Resource referenced in Fn::Sub expression with logical ID ...
  return ...
}
```